File: web/src/api/carpoolClient.test.ts
import { describe, expect, it, vi } from "vitest"

import { CarpoolClient } from "@/api/carpoolClient"
import type { CarpoolSpace, CarpoolSummary } from "@/api/types"

const sampleSpace: CarpoolSpace = {
  id: "s1",
  name: "Soccer",
  membership: "OWNER",
  inviteCode: "AB12CD34",
  callerFeedId: "f1",
  members: [
    { circleId: "c1", circleName: "House A", membership: "OWNER" },
  ],
  pendingRequests: [],
}

const sampleSummary: CarpoolSummary = {
  circleRole: "ORGANIZER",
  feeds: [
    {
      feedId: "f1",
      feedName: "Soccer",
      status: "NONE",
      spaceId: null,
      spaceName: null,
    },
  ],
  spaces: [],
}

describe("CarpoolClient", () => {
  it("loads summary with Bearer auth", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify(sampleSummary), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      }),
    )
    const client = new CarpoolClient("http://localhost:8080", fetchFn)

    const summary = await client.getSummary("tok")

    expect(summary.circleRole).toBe("ORGANIZER")
    expect(summary.feeds[0]?.status).toBe("NONE")
    const [url, init] = fetchFn.mock.calls[0] as [string, RequestInit]
    expect(url).toBe("http://localhost:8080/api/carpool")
    expect(init.headers).toMatchObject({ Authorization: "Bearer tok" })
  })

  it("enables, joins, requests, admits, declines, regenerates, and leaves", async () => {
    const fetchFn = vi
      .fn()
      .mockResolvedValueOnce(
        new Response(JSON.stringify(sampleSpace), {
          status: 201,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(JSON.stringify({ ...sampleSpace, membership: "MEMBER" }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(JSON.stringify(sampleSpace), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(
          JSON.stringify({
            id: "r1",
            spaceId: "s1",
            circleId: "c2",
            circleName: "House B",
            requestedByAdultId: "a2",
            requestedByDisplayName: "Sam",
          }),
          { status: 201, headers: { "Content-Type": "application/json" } },
        ),
      )
      .mockResolvedValueOnce(
        new Response(JSON.stringify(sampleSpace), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(new Response(null, { status: 204 }))
      .mockResolvedValueOnce(
        new Response(JSON.stringify({ code: "XY98ZW76" }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(new Response(null, { status: 204 }))

    const client = new CarpoolClient("http://localhost:8080", fetchFn)

    await client.enable("tok", "f1")
    await client.join("tok", "AB12CD34")
    await client.getSpace("tok", "s1")
    await client.createRequest("tok", "s1")
    await client.admit("tok", "s1", "r1")
    await client.decline("tok", "s1", "r1")
    const invite = await client.regenerateInvite("tok", "s1")
    await client.leave("tok", "s1")

    expect(invite.code).toBe("XY98ZW76")
    const urls = fetchFn.mock.calls.map((call) => (call as [string, RequestInit])[0])
    expect(urls).toEqual([
      "http://localhost:8080/api/carpool/enable",
      "http://localhost:8080/api/carpool/join",
      "http://localhost:8080/api/carpool/spaces/s1",
      "http://localhost:8080/api/carpool/spaces/s1/requests",
      "http://localhost:8080/api/carpool/spaces/s1/requests/r1/admit",
      "http://localhost:8080/api/carpool/spaces/s1/requests/r1/decline",
      "http://localhost:8080/api/carpool/spaces/s1/invite/regenerate",
      "http://localhost:8080/api/carpool/spaces/s1/leave",
    ])
    expect((fetchFn.mock.calls[0] as [string, RequestInit])[1].method).toBe("POST")
    expect((fetchFn.mock.calls[0] as [string, RequestInit])[1].body).toBe(
      JSON.stringify({ feedId: "f1" }),
    )
    expect((fetchFn.mock.calls[1] as [string, RequestInit])[1].body).toBe(
      JSON.stringify({ code: "AB12CD34" }),
    )
  })

  it("surfaces server error messages", async () => {
    const fetchFn = vi.fn().mockResolvedValue(
      new Response(JSON.stringify({ message: "Organizer role required" }), {
        status: 403,
        headers: { "Content-Type": "application/json" },
      }),
    )
    const client = new CarpoolClient("http://localhost:8080", fetchFn)

    await expect(client.enable("tok", "f1")).rejects.toThrow("Organizer role required")
  })

  it("lists, creates, accepts, passes, cancels, and withdraws rides", async () => {
    const ride = {
      id: "ride-1",
      spaceId: "s1",
      eventKey: "UID:practice",
      requestingCircleId: "c2",
      requestingCircleName: "House B",
      requestedByAdultId: "a2",
      kidIds: ["k1"],
      kidFirstNames: ["Mia"],
      seats: 1,
      pickupPlaceName: "Home",
      pickupAddress: "1 Main St",
pickupTown: null,
detourMinutes: null,
      status: "PENDING",
      passedByMe: false,
      passedByAdultNames: [],
      acceptedByAdultId: null,
      acceptingCircleId: null,
      acceptingCircleName: null,
      vehicleId: null,
      vehicleLabel: null,
    }
    const event = {
      eventKey: "UID:practice",
      title: "Practice",
      startsAt: "2026-08-21T16:00:00Z",
      endsAt: null,
      defaultKidIds: ["k1"],
      ownRequest: null,
      otherRequests: [ride],
    }
    const fetchFn = vi
      .fn()
      .mockResolvedValueOnce(
        new Response(JSON.stringify([event]), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(JSON.stringify(ride), {
          status: 201,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(JSON.stringify({ ...ride, status: "ACCEPTED", vehicleId: "v1" }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(JSON.stringify({ ...ride, passedByMe: true, passedByAdultNames: ["Alex"] }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(JSON.stringify({ ...ride, status: "CANCELLED" }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
      .mockResolvedValueOnce(
        new Response(JSON.stringify({ ...ride, status: "PENDING" }), {
          status: 200,
          headers: { "Content-Type": "application/json" },
        }),
      )
    const client = new CarpoolClient("http://localhost:8080", fetchFn)

    await expect(
      client.listRides("tok", "s1", "2026-08-01T00:00:00Z", "2026-08-31T00:00:00Z"),
    ).resolves.toMatchObject([{ title: "Practice" }])
    await expect(
      client.createRide("tok", "s1", { eventKey: "UID:practice" }),
    ).resolves.toMatchObject({ id: "ride-1", passedByMe: false })
    await expect(
      client.acceptRide("tok", "s1", "ride-1", { vehicleId: "v1" }),
    ).resolves.toMatchObject({ status: "ACCEPTED" })
    await expect(client.passRide("tok", "s1", "ride-1")).resolves.toMatchObject({
      passedByMe: true,
      passedByAdultNames: ["Alex"],
      status: "PENDING",
    })
    await expect(client.cancelRide("tok", "s1", "ride-1")).resolves.toMatchObject({
      status: "CANCELLED",
    })
    await expect(client.withdrawRide("tok", "s1", "ride-1")).resolves.toMatchObject({
      status: "PENDING",
    })

    const urls = fetchFn.mock.calls.map((call) => (call as [string, RequestInit])[0])
    expect(urls).toEqual([
      "http://localhost:8080/api/carpool/spaces/s1/rides?from=2026-08-01T00%3A00%3A00Z&to=2026-08-31T00%3A00%3A00Z",
      "http://localhost:8080/api/carpool/spaces/s1/rides",
      "http://localhost:8080/api/carpool/spaces/s1/rides/ride-1/accept",
      "http://localhost:8080/api/carpool/spaces/s1/rides/ride-1/pass",
      "http://localhost:8080/api/carpool/spaces/s1/rides/ride-1/cancel",
      "http://localhost:8080/api/carpool/spaces/s1/rides/ride-1/withdraw",
    ])
    expect((fetchFn.mock.calls[0] as [string, RequestInit])[1].headers).toMatchObject({
      Authorization: "Bearer tok",
    })
    expect((fetchFn.mock.calls[1] as [string, RequestInit])[1].body).toBe(
      JSON.stringify({ eventKey: "UID:practice" }),
    )
    expect((fetchFn.mock.calls[2] as [string, RequestInit])[1].body).toBe(
      JSON.stringify({ vehicleId: "v1" }),
    )
    expect((fetchFn.mock.calls[3] as [string, RequestInit])[1].method).toBe("POST")
    expect((fetchFn.mock.calls[3] as [string, RequestInit])[1].body).toBeUndefined()
  })
})

File: mobile/sharedLogic/src/commonTest/kotlin/org/example/project/CarpoolClientTest.kt
package org.example.project

import io.ktor.client.HttpClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.http.headersOf
import io.ktor.serialization.kotlinx.json.json
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.Json

class CarpoolClientTest {

    @Test
    fun getSummarySendsBearerAndDecodesFeedStatus() =
        runTest {
            val mockEngine =
                MockEngine { request ->
                    assertEquals("/api/carpool", request.url.encodedPath)
                    assertEquals(HttpMethod.Get, request.method)
                    assertEquals("Bearer tok", request.headers[HttpHeaders.Authorization])
                    respond(
                        content =
                            """{"circleRole":"ORGANIZER","feeds":[{"feedId":"f1","feedName":"Soccer","status":"NONE","spaceId":null,"spaceName":null}],"spaces":[]}""",
                        status = HttpStatusCode.OK,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }

            val client = CarpoolClient("http://localhost:8080", mockHttpClient(mockEngine))
            val summary = client.getSummary("tok")

            assertEquals(FamilyRole.ORGANIZER, summary.circleRole)
            assertEquals(CarpoolFeedStatusKind.NONE, summary.feeds.single().status)
            assertNull(summary.feeds.single().spaceId)
            assertTrue(summary.spaces.isEmpty())
        }

    @Test
    fun enableJoinRequestAdmitDeclineRegenerateAndLeave() =
        runTest {
            val spaceJson =
                """{"id":"s1","name":"Soccer","membership":"OWNER","inviteCode":"AB12CD34","callerFeedId":"f1","members":[{"circleId":"c1","circleName":"House A","membership":"OWNER"}],"pendingRequests":[]}"""
            val mockEngine =
                MockEngine { request ->
                    when {
                        request.url.encodedPath == "/api/carpool/enable" &&
                            request.method == HttpMethod.Post -> {
                            assertEquals("Bearer tok", request.headers[HttpHeaders.Authorization])
                            respond(
                                content = spaceJson,
                                status = HttpStatusCode.Created,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        }
                        request.url.encodedPath == "/api/carpool/join" &&
                            request.method == HttpMethod.Post ->
                            respond(
                                content =
                                    """{"id":"s1","name":"Soccer","membership":"MEMBER","inviteCode":"AB12CD34","callerFeedId":"f1","members":[{"circleId":"c1","circleName":"House A","membership":"OWNER"},{"circleId":"c2","circleName":"House B","membership":"MEMBER"}],"pendingRequests":[]}""",
                                status = HttpStatusCode.OK,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        request.url.encodedPath == "/api/carpool/spaces/s1" &&
                            request.method == HttpMethod.Get ->
                            respond(
                                content = spaceJson,
                                status = HttpStatusCode.OK,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        request.url.encodedPath == "/api/carpool/spaces/s1/requests" &&
                            request.method == HttpMethod.Post ->
                            respond(
                                content =
                                    """{"id":"r1","spaceId":"s1","circleId":"c2","circleName":"House B","requestedByAdultId":"a2","requestedByDisplayName":"Sam"}""",
                                status = HttpStatusCode.Created,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        request.url.encodedPath == "/api/carpool/spaces/s1/requests/r1/admit" &&
                            request.method == HttpMethod.Post ->
                            respond(
                                content = spaceJson,
                                status = HttpStatusCode.OK,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        request.url.encodedPath == "/api/carpool/spaces/s1/requests/r1/decline" &&
                            request.method == HttpMethod.Post ->
                            respond(content = "", status = HttpStatusCode.NoContent)
                        request.url.encodedPath == "/api/carpool/spaces/s1/invite/regenerate" &&
                            request.method == HttpMethod.Post ->
                            respond(
                                content = """{"code":"XY98ZW76"}""",
                                status = HttpStatusCode.OK,
                                headers = headersOf(HttpHeaders.ContentType, "application/json"),
                            )
                        request.url.encodedPath == "/api/carpool/spaces/s1/leave" &&
                            request.method == HttpMethod.Post ->
                            respond(content = "", status = HttpStatusCode.NoContent)
                        else -> error("Unexpected ${request.method} ${request.url.encodedPath}")
                    }
                }

            val client = CarpoolClient("http://localhost:8080", mockHttpClient(mockEngine))

            val enabled = client.enable("tok", "f1")
            assertEquals(CarpoolSpaceMembership.OWNER, enabled.membership)
            assertEquals("AB12CD34", enabled.inviteCode)
            assertEquals("House A", enabled.members.single().circleName)

            val joined = client.join("tok", "AB12CD34")
            assertEquals(CarpoolSpaceMembership.MEMBER, joined.membership)

            assertEquals("Soccer", client.getSpace("tok", "s1").name)

            val request = client.createRequest("tok", "s1")
            assertEquals("r1", request.id)
            assertEquals("House B · requested by Sam", request.displayLabel())

            assertEquals("s1", client.admit("tok", "s1", "r1").id)
            client.decline("tok", "s1", "r1")
            assertEquals("XY98ZW76", client.regenerateInvite("tok", "s1").code)
            client.leave("tok", "s1")
        }

    @Test
    fun enableSurfacesOrganizerForbiddenMessage() =
        runTest {
            val mockEngine =
                MockEngine {
                    respond(
                        content = """{"message":"Organizer role required"}""",
                        status = HttpStatusCode.Forbidden,
                        headers = headersOf(HttpHeaders.ContentType, "application/json"),
                    )
                }
            val client = CarpoolClient("http://localhost:8080", mockHttpClient(mockEngine))
            val error =
                assertFailsWith<AuthApiException> {
                    client.enable("tok", "f1")
                }
            assertEquals("Organizer role required", error.message)
        }
}

class CarpoolModelsTest {

    @Test
    fun circleDisplayNameFallsBackToYourFamily() {
        assertEquals("Your family", circleDisplayName(null))
        assertEquals("Your family", circleDisplayName("  "))
        assertEquals("House A", circleDisplayName("House A"))
    }

    @Test
    fun feedStatusLabels() {
        assertEquals("No carpool", CarpoolFeedStatusKind.NONE.statusLabel())
        assertEquals("Carpool available", CarpoolFeedStatusKind.AVAILABLE.statusLabel())
        assertEquals("Requested", CarpoolFeedStatusKind.REQUESTED.statusLabel())
        assertEquals("Member", CarpoolFeedStatusKind.MEMBER.statusLabel())
        assertEquals("Owned", CarpoolFeedStatusKind.OWNER.statusLabel())
    }

    @Test
    fun enableConfirmAsksFamilyToOwnTheSpace() {
        assertTrue(
            enableCarpoolConfirmMessage("Soccer").contains("own the carpool for Soccer"),
        )
    }

    @Test
    fun caregiverEmptyHintDoesNotMentionFeeds() {
        val caregiver =
            CarpoolSummary(circleRole = FamilyRole.CAREGIVER, feeds = emptyList(), spaces = emptyList())
        val organizer =
            CarpoolSummary(circleRole = FamilyRole.ORGANIZER, feeds = emptyList(), spaces = emptyList())
        assertTrue(caregiver.hasNoCarpools())
        assertFalse(caregiver.emptyHint().contains("Feeds"))
        assertTrue(organizer.emptyHint().contains("Feeds"))
    }

    @Test
    fun primaryActionHidesEnableForCaregiver() {
        val none =
            CarpoolFeedStatus(
                feedId = "f1",
                feedName = "Soccer",
                status = CarpoolFeedStatusKind.NONE,
            )
        assertEquals(CarpoolPrimaryAction.ENABLE, none.primaryAction(FamilyRole.ORGANIZER))
        assertEquals(CarpoolPrimaryAction.NONE, none.primaryAction(FamilyRole.CAREGIVER))
        val available =
            none.copy(status = CarpoolFeedStatusKind.AVAILABLE, spaceId = "s1", spaceName = "Soccer")
        assertEquals(CarpoolPrimaryAction.REQUEST, available.primaryAction(FamilyRole.CAREGIVER))
    }
}

private fun mockHttpClient(engine: MockEngine): HttpClient =
    HttpClient(engine) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true })
        }
    }
